cobalt/black_box_tests/testdata/scroll.html

```diff
@@ -35,18 +34,109 @@
 <body>
   <div class="app" style="overflow:auto;">
     <div id="row" class="row" style="height: 20rem;">
-      <div id="top_one" class="tile">One</div>
+      <div id="tile" class="tile">One</div>
     </div>
   </div>
 </body>
 <script>
+  // Fail if the test is not finished within 15 seconds.
+  var kTimeout = 30 * 1000;
+  var failTimer = setTimeout(fail, kTimeout);
+
+  function fail() {
+    console.log("Failing due to timeout!");
+    assertTrue(false);
+  }
+
+  function phasename(phase) {
+    switch (phase) {
+      case 0: return 'none';
+      case 1: return 'capturing';
+      case 2: return 'at target';
+      case 3: return 'bubbling';
+    }
+    return ' [unknown: ' + phase + ']';
+  }
+
+  const expected_events = [
+    // name, id, phase
+    // actions.move_to_element(row).pause(_SLEEP_AFTER_MOVE_TIME)
+    ['pointermove', 'tile', 'at target'],
+    ['pointermove', 'row', 'bubbling'],
+
+    // actions.click_and_hold(row).pause(_SLEEP_AFTER_MOVE_TIME)
+    ['pointermove', 'tile', 'at target'],
+    ['pointermove', 'row', 'bubbling'],
+
+    // actions.move_by_offset(-500, 0).pause(_SLEEP_AFTER_MOVE_TIME)
+    ['pointerdown', 'tile', 'at target'],
+    ['pointerdown', 'row', 'bubbling'],
+    ['pointercancel', 'tile', 'at target'],
+    ['pointercancel', 'row', 'bubbling'],
+    ['scroll', 'row', 'at target'],
+  ]
+
+  let current_event_number = 0
+  let failure_count = 0;
+
+  function CheckEvent(name, id, phase) {
+    while ((current_event_number < expected_events.length) &&
+      ((expected_events[current_event_number][0] != name) ||
```

Comment:
I think we could just pass CheckEvent an object that matches the type of each item in expected_events and compare that directly instead of comparing each field

Reply:
Well, I'm not sure we can do this. These two are not reference equal, so directly using "==" to compare this two objects will always get false?

Reply:
I made the code change here, but I think we still need to compare the value of each field, no?

Reply:
Yes, you're totally right. I refactored checkEvent slightly to destructure and use template literals, lmk if you think this is okay

```
function checkEvent(e) {
  while (current_event_number < expected_events.length) {
    current_event_number += 1;

    const {name, id, phase} = expected_events[current_event_number]
    if (name === e.name && id === e.id && phase === e.phase) {
      break;
    }
    console.error(`Missing Event ['${name}', '${id}', '${phase}']`);
    failure_count += 1;
    current_event_number += 1;
  }
  if (current_event_number >= expected_events.length) {
    console.error(`Unexpected Event ['${e.name}', '${e.id}', '${e.phase}']`);
    failure_count += 1;
  }
}
```

let's also try to use template literals below

Reply:
That looks great. But prefer to address this in the stacked PR here: https://github.com/youtube/cobalt/pull/1702 to avoid rebase work. Also this function is mostly copied from pointer_tests.html, in PR#1702, I put everything can be used in a util js file, so it actually make sense to put the refactor work there.

Reply:
Sure, doing it in a follow up works for me

Reply:
@andrewsavage1  Well the above refactor actually have bugs, and make the logic more complicated. 
1) when the while loop get `current_event_number == expected_events.length`, and it will fall into the below `if (current_event_number >= expected_events.length)` check and through error.
2) In the first line of the while loop, `current_event_number += 1;` resulting missing expected_events[0]. (This can be fixed easily.)
3) For the missing event case, we plus `current_event_number += 1;` twice.

I tried to fix the above bugs following the proposal refactor way, but it's actually very tricky to find the right place to put `current_event_number += 1;` and handle the case `current_event_number == expected_events.length`. And in the end, it makes the logic less readable then the original method. 

I'm going to change this back in the following PR, unless you find another clear way to write this function.

Reply:
I think this fixes everything, though not sure what you mean by (1)
```
function checkEvent(e) {
  while (current_event_number < expected_events.length) {
    const {name, id, phase} = expected_events[current_event_number]
    current_event_number += 1;
    if (name === e.name && id === e.id && phase === e.phase) {
      break;
    }

    console.error(`Missing Event ['${name}', '${id}', '${phase}']`);
    failure_count += 1;
  }
  if (current_event_number >= expected_events.length) {
    console.error(`Unexpected Event ['${e.name}', '${e.id}', '${e.phase}']`);
    failure_count += 1;
  }
}
```

Reply:
Nope, this one still have issue 1).

What I mean is, assume all expected_events are correct, when we check the last expected_events, `current_event_number == expected_events.length -1`, and in the while loop, current_event_number++, then `current_event_number == expected_events.length `-> after the code break from the loop, it meet the condition `if (current_event_number >= expected_events.length)`, so `failure_count +=1`, and that's not right.

So in this way, all checks will fail when checkEvent(last_event).

Reply:
That would just mean doing
```
function checkEvent(e) {
  let matched = false;
  while (current_event_number < expected_events.length) {
    const {name, id, phase} = expected_events[current_event_number]
    current_event_number++;
    if (name === e.name && id === e.id && phase === e.phase) {
      matched = true;
      break;
    }

    console.error(`Missing Event ['${name}', '${id}', '${phase}']`);
    failure_count++;
  }
  if (!matched) {
    console.error(`Unexpected Event ['${e.name}', '${e.id}', '${e.phase}']`);
    failure_count++;
  }
}
```
which I still think is pretty clean. if you disagree, my main issue is with the large while loop check, so I'd be fine if we did the check in two steps or made a helper function for checking object equality